sdk/src/traits.rs

```diff
@@ -0,0 +1,110 @@
+use serde::{de::DeserializeOwned, Serialize};
+use std::fs;
+use std::path::{Path, PathBuf};
+
+use crate::compile::*;
+
+/// A compute resource.
+pub trait Compute {}
+
+/// Indicator type that local compute will be used for proving the zkVM.
+pub enum Local {}
+impl Compute for Local {}
+
+/// A view capturing the output of a zkVM execution.
+pub struct View<U: DeserializeOwned> {
```

Comment:
I debated this interface initially. In general, I think I pretty strongly prefer the single specification of the input/output interface at run. However, as you note, there are plenty of knock-on effects for having that. 

I'll add one more though, that I think just caused me to agree with you that this sort of approach is the better way to go. It's possible that the host won't necessarily know the appropriately `U`, since different computational paths of the guest program might write different output types to the tape. If using the single interface, the host would have to rerun the entire computation to "test" a different `U`, whereas using this interface they can invoke `output` multiple times with different `U` until one doesn't fail.

Reply:
This is what my comment says: "A simple use case is serializing different types based on the input. "

Reply:
Ah sorry, didn't process that.